1단계 미션 제출합니다

Eungu-Kwon opened the pull request.

kriery commented on src/main/java/context/ApplicationContext.java:
```diff
@@ -2,20 +2,21 @@
+                new UserCreateController(),
+                new UserLoginController(),
+                new UserListController())
+                .collect(Collectors.toMap(Controller::getPath, x -> x));
```
`x -> x` 대신에 `Function.identity()` 라는 것도 있긴 합니다 (하지만 `x -> x` 이 더 짧죠 ^^)

humbledude replied:
네네 맞아요 저도 그거랑 비교해봤는데 ㅎㅎㅎ
Function.identyty() 는 x->x 를 리턴하더군요....
저는 x -> x 가 더 보기 좋다고 판단되어 이거를 썼는데,
성능은 뭐 영향없으니 취향껏 고르면 될꺼 같네요~ ㅎㅎ

kriery commented on src/main/java/controller/Controller.java:
```diff
@@ -15,6 +15,8 @@ public HttpResponse handleRequest(HttpRequest httpRequest) {
         }
     }
 
+    public abstract String getPath();
```
controllerMap을 자동으로 구성하기 위한 공통 method 👍

kriery commented on src/main/java/webserver/http/ParameterValidator.java:
```diff
@@ -0,0 +1,17 @@
+            throw new IllegalArgumentException(absentParameters + " 가 꼭 필요해요");
+        }
+    }
+}
```
새로운 helper 가 출현했네요 ❤️

kriery commented on src/main/java/controller/UserCreateController.java:
```diff
@@ -1,42 +1,39 @@
+        String userId = request.getParameter("userId");
+        String password = request.getParameter("password");
+        String name = request.getParameter("name");
+        String email = request.getParameter("email");
```
controller에서 request에서 받아온 raw data를 직접 service로 넘기기 보다는,
model인 `User`로 감싸서 넘기는것은 어떨까요?

humbledude replied:
네네 좋습니다 ㅎㅎ
제가 User 라는 도메인 객체는 Controller 에서 만들어야 되나 Service 에서 만들어야 되나 고민을 하던 차였습니다~
의견 감사합니다 ㅎㅎ

kriery reviewed (approved):
step1 에서 클래스별 책임 분리를 너무 잘 해 주셔서 step2가 금방 진행되셨네요 💯 
controler / service 간에 데이터 이동에 대해서 간략한 피드백을 드렸어요,
취향을 타는 문제일 수 있어서, 수정 요청으로 드리진 않았어요.
확인해보시고 합리적은 방법으로 진행하시면 될것 같습니다. 👍
그럼 이번 미션 마지막 단계로 go-go-go!

seungy0 committed:
feat: GET 요청을 받을 때 요청한 페이지를 응답하는 기능 구현

seungy0 committed:
feat: 요청에 맞는 Content-Type 적용

stylesheet를 포함한 static 디렉토리 내부의 파일에 대해 적절하게 응답하도록 변경

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: 회원가입을 위해 query string 파싱

요청 데이터를 담는 HttpRequest와 응답을 처리하는 HttpResponse 클래스 작성

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: post 방식으로 요청을 보내는 회원가입 페이지 구현

- 헤더의 Content-Length만큼 Request에서 body를 읽어오도록 구현
- PostRequestHandler에서 요청에 대한 로직 처리

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: 회원가입 후 index.html로 redirect

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seokhong commented on src/main/java/utils/IOUtils.java:
`HttpRequestTestHeader` 테스트가 실패하는 것으로 보이네요. 테스트가 성공할 수 있도록 구현 코드를 조금 수정해 보시면 좋겠습니다. :)

seokhong commented on src/main/java/utils/IOUtils.java:
```diff
@@ -17,4 +20,30 @@ public static String readData(BufferedReader br, int contentLength) throws IOExc
         return String.copyValueOf(body);
     }
+
+    public static String readRequestLine(BufferedReader reader) throws IOException {
```
`readRequestLine`, `readHeader`는 특정한 조건 하에서만 의도한대로 동작할 것으로 보입니다.
이 기능은 `RequestHandler`에 구현해보면 어떨까요? `IOUtils`는 범용적인 기능을 가지는 유틸리티 클래스로 보이는데, 이 두 메서드는 정확히 첫 줄 혹은 헤더가 시작되는 줄에 호출되어야 의도한대로 동작할 것으로 보이네요.

seokhong commented on src/main/java/webserver/GetRequestHandler.java:
```diff
@@ -0,0 +1,33 @@
+    private static final Logger logger = LoggerFactory.getLogger(GetRequestHandler.class);
+
+    public static void getHandler(HttpRequest httpRequest, OutputStream out) throws IOException {
+        if (httpRequest.getPath().equals("/user/create")) {
```
사용자를 생성하는 기능은 `PostRequestHandler`로 변경했으니, 이 코드는 제거해보면 어떨까요?

seokhong commented on src/main/java/webserver/HttpResponse.java:
```diff
@@ -0,0 +1,44 @@
+        this.body = body;
+    }
+
+    public void sendResponse(OutputStream out)
```
`HttpResponse`는 응답을 담는 객체로 보이네요. 현재와 같은 구조에서는 응답을 전달하기 위해, 모든 핸들러에서 `OutputStream`을 전달 받아서 `HttpResponse`에게 전달해줘야 할 것으로 보입니다.
각 핸들러에서는 `HttpResponse`를 만들고, 클라이언트에게 전달하는 것은 `RequestHandler`에서 해보면 어떨까요?

seokhong replied:
`HttpResponse` 에게 `OutputStream`을 전달하지 말고, `RequestHandler`에서 `DataOutputStream`을 생성해서 `HttpResponse`를 write 해보면 어떨까요?
`HttpResponse`는 응답 값을 가지고 있는 객체이고, `OutputStream`은 값을 출력하는 장치입니다. 현재의 코드는 마치 `종이`에게 `프린터`를 전달해서 출력하라고 시키는 것처럼 보이는 것 같은데, 어떻게 생각하시나요?

seokhong commented on src/main/java/webserver/HttpResponse.java:
```diff
@@ -0,0 +1,44 @@
+
+    public void sendResponse(OutputStream out)
+        throws IOException {
+        DataOutputStream dos = new DataOutputStream(out);
```
`DataOutputStream`도 close 가 필요한 자원으로 보이는데, try-with-resources 로 사용하지 않으신 이유가 있을까요? :)

seokhong commented on src/main/java/webserver/PostRequestHandler.java:
```diff
@@ -0,0 +1,36 @@
+    public static void handler(HttpRequest httpRequest, OutputStream out) throws IOException {
+        if (httpRequest.getPath().equals("/user/create")) {
+            createUser(httpRequest.getForm());
+            HttpResponse httpResponse = new HttpResponse("302", "Found",
```
302, 404 등은 상태를 나타내는 HTTP status code 일텐데요. 상태 메시지와 코드를 별도의 객체로 관리해서, 무슨 상태가 있는지 볼 수 있도록 하고, 오타 등의 실수도 방지해보면 어떨까요?

seokhong commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,75 @@
+            byte[] body = Files.readAllBytes(file.toPath());
+            String contentType = Files.probeContentType(file.toPath());
+            HttpResponse httpResponse = new HttpResponse("200", "OK",
+                Map.of("Content-Type", contentType != null ? contentType : "null"),
```
의미를 갖는 문자열은 enum, 상수 등을 적절하게 활용해 보시면 좋겠습니다.

seokhong commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+        String[] tokens = input.split(delimiter);
+        Map<String, String> map = new HashMap<>();
+        for (String token : tokens) {
+            String[] keyValue = token.split("=");
```
`=`가 우리가 기대한 형식과 다르게 들어오면 어떻게 될까요?
예를 들면, `params=a=b`와 같이요.

Eungu-Kwon replied:
안녕하세요!
제가 테스트 해보았을 때는 구분자로 쓰이는 = 외에는 %3D로 인코딩 되어서 저장되는 것을 확인했습니다.
혹시 =이나 다른 구분자를 따로 처리하는 로직이 필요한지 궁금합니다!

seokhong replied:
서버를 띄워서 테스트했을 때, 의도와 다르게 파라메터가 처리되는 것으로 보이네요!
<img width="1376" alt="스크린샷 2024-04-11 오후 6 36 34" src="https://github.com/next-step/jwp-was-kakao/assets/6284253/f75e03b7-12ac-4fc4-952e-267288d0ea30">
<img width="279" alt="스크린샷 2024-04-11 오후 6 37 00" src="https://github.com/next-step/jwp-was-kakao/assets/6284253/e735b525-afa5-4766-acc4-98c881fc7f68">

seokhong commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,75 @@
+                headers.get("Content-Length") != null ? headers.get("Content-Length") : "0"));
+            HttpRequest httpRequest = new HttpRequest(requestLine, headers, body);
+
+            if (httpRequest.isMethod("GET") && responseResources(httpRequest.getPath(), out)) {
```
정적 리소스에 대한 처리도 `GET` 요청에 대한 처리이니 분리할 수 있지 않을까요?

seokhong replied:
현재 구현에서는 http method 가 늘어날 때마다 분기가 늘어날 것으로 보이는데요. 개선할 수 있는 방법을 고민해보면 어떨까요?

seokhong reviewed (changes requested):
은구님, 1단계 미션 잘 구현하셨습니다. 👍
전체적으로 기능 잘 동작하고 테스트 성공하는 것 확인했습니다.
코멘트 확인 부탁드립니다. 🙂

Eungu-Kwon committed:
refactor: 피드백 받은 사항 적용

Co-authored-by: seungy0 <dltmdrb98@gmail.com>

seungy0 committed:
refactor: query -> queryParam으로 프로퍼티명 변경

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seokhong commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,79 @@
+                headers.get(HttpHeaders.CONTENT_LENGTH) != null ? headers.get(HttpHeaders.CONTENT_LENGTH) : "0"));
+            HttpRequest httpRequest = new HttpRequest(requestLine, headers, body);
+
+            HttpResponse http404Response = new HttpResponse(HttpStatus.NOT_FOUND, null, null);
```
이 값은 handler를 찾지 못했을 때만 필요한 것으로 보이네요.
별도의 메서드로 분리해서 `orElse` 메서드에서 호출해보면 어떨까요? 코드를 읽을 때 중간에 404 응답을 왜 만드는지 궁금한 순간이 있었네요!

seokhong commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,79 @@
+
+            HttpResponse http404Response = new HttpResponse(HttpStatus.NOT_FOUND, null, null);
+
+            Map<String, MethodRequestHandler> handlers = new HashMap<>();
```
핸들러 맵, 핸들러는 서버가 뜬 이후에 변경되지 않을 값으로 보이네요.
필드 등으로 미리 만들어 놓고 재사용해보면 어떨까요?

seokhong commented on src/main/java/webserver/http/HttpStatus.java:
```diff
@@ -0,0 +1,23 @@
+	}
+
+	public static final String HTTP_VERSION = "HTTP/1.1";
+}
```
파일 마지막에 newline 이 없다고 깃헙에서 경고를 띄워주네요.  
파일 마지막의 newline 은 무슨 역할을 하는지 학습해보신 후, IDE 설정을 통해 휴먼 에러를 방지해보시면 좋겠습니다.

seokhong commented on src/main/java/webserver/MethodRequestHandler.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.Optional;
+
+public interface MethodRequestHandler {
+	Optional<HttpResponse> handler(HttpRequest httpRequest) throws IOException;
```
메서드의 이름은 `handle` 이 더 적절할 것으로 보이네요!

seokhong reviewed (changes requested):
은구님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
어제 제가 코멘트에 은구님 이름을 잘못 작성하는 실수를 했네요. 죄송합니다. 🙇 
조금 더 고민해보시면 좋을 것 같은 내용이 있어 코멘트 추가로 남겼습니다. 확인 부탁드립니다. 🙂

seungy0 committed:
refactor: 인덱스 상수 지정, 쿼리파라미터 trim

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
refactor: 입력스트림에 null이 들어올 때에 대한 처리

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
refactor: 쿼리파라미터에 키만 들어왔을 때, 키 밸류가 들어오지 않았을 때에 대한 처리

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
style: 코드 스타일 적용

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

Eungu-Kwon committed:
refactor: 피드백 받은 사항 적용
- 쿼리 파라미터에 = 문자열이 있는 경우 처리
- RequestHandler에서 HttpResponse를 받아 직접 요청에 응답

Co-authored-by: seungy0 <dltmdrb98@gmail.com>

seokhong commented on src/test/java/webserver/HttpRequestTest.java:
```diff
@@ -14,4 +14,11 @@ void request_resttemplate() {
     }
+
+    @Test
+    void request_resttemplate2() {
```
위의 `request_resttemplate`과 동일한 테스트 코드로 보이네요. 중복되는 코드는 제거하면 어떨까요?

seokhong commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class HttpRequest {
```
`HttpRequest`의 구성 요소인 메서드, 경로, 헤더 등을 감싸는 객체를 만들어 보는 건 어떨까요?
[http message 형식](https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages)을 보면 request line, headers, body 와 같이 명확한 이름과 형식이 정의되어 있으니, 클래스로 구현하면 가독성이 더 좋아질 것 같아요!

seokhong commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,81 @@
+        return body == null ? new HashMap<>() : parseKeyValuePairs(body, "&");
+    }
+
+    private Map<String, String> parseKeyValuePairs(String input, String delimiter) {
```
`delimiter`는 항상 `&`가 들어오니, 이 값을 상수로 정의하고 사용해도 괜찮지 않을까요?

seokhong commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,81 @@
+                .map(String::trim)
+                .collect(Collectors.toList());
+
+            Optional.of(keyValue)
```
`Optional`로 감싸고 filter, ifPresent를 활용하는 방법도 괜찮지만, 때로는 if 문을 이용하는 코드가 더 간결하고 가독성이 좋기도 합니다.
`keyValue`는 절대 null이 될 수 없는 값으로 보이는데, 코드를 읽는 사람에게는 어떤 변수를 `Optional`로 감싸는 순간 null 이 될 수도 있다는 의미로 전달될 수도 있어 보입니다.

seokhong commented on src/main/java/webserver/http/HttpMethods.java:
```diff
@@ -0,0 +1,9 @@
+package webserver.http;
+
+public class HttpMethods {
```
문자열 상수로 관리하는 것과 enum 으로 관리하는 것에도 차이가 있을 것 같아요.
문자열 상수와 enum의 장단점 등을 고민해보셔도 좋겠습니다. :)

seokhong commented on src/main/java/webserver/http/HttpHeaders.java:
```diff
@@ -0,0 +1,8 @@
+package webserver.http;
```
http 패키지는 마치 상수만 정의하기 위한 패키지처럼 보이기도 하네요.
webserver 패키지의 일부가 http 스펙이라면 http 패키지로 옮기는 것이 적절하지는 않을까요?
반대로 webserver 에서 사용하는 상수를 http 패키지에 정의한 것이라면, 이 두 패키지는 결합도가 강하니 한 패키지로 관리하거나, http 패키지가 webserver 에서 사용하는 상수임을 나타내는 패키지로 변경하는게 가독성이 더 좋지 않을까 싶기도 합니다. :)

seokhong reviewed (approved):
은구님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
1단계 PR 머지하겠습니다.
추가로 코멘트 남긴 내용은 2단계 미션을 진행하시며 참고해 보시면 좋겠습니다.
2단계 미션에서 뵙겠습니다. 🙂
